// mcpcac/dist/oauth-provider.js
/**
 * File-based OAuth provider implementation for CLI usage.
 * Implements the OAuthClientProvider interface from MCP SDK.
 *
 * Unlike server-based implementations that use a database,
 * this stores state in memory and calls onStateUpdated for persistence.
 */
export class FileOAuthProvider {
    _clientInformation;
    _codeVerifier;
    _tokens;
    _redirectStartAuthUrl;
    serverUrl;
    redirectUri;
    clientName;
    onStateUpdated;
    constructor(options) {
        this.serverUrl = options.serverUrl;
        this.redirectUri = options.redirectUri;
        this.clientName = options.clientName;
        this._tokens = options.tokens;
        this._clientInformation = options.clientInformation;
        this._codeVerifier = options.codeVerifier;
        this.onStateUpdated = options.onStateUpdated;
    }
    get redirectUrl() {
        return this.redirectUri;
    }
    /**
     * The authorization URL to redirect the user to.
     * Set by redirectToAuthorization().
     */
    get redirectStartAuthUrl() {
        return this._redirectStartAuthUrl;
    }
    async clientInformation() {
        return this._clientInformation;
    }
    async saveClientInformation(clientInformation) {
        this._clientInformation = clientInformation;
        this.notifyStateUpdated();
    }
    async codeVerifier() {
        if (!this._codeVerifier) {
            throw new Error("Code verifier not set");
        }
        return this._codeVerifier;
    }
    async saveCodeVerifier(codeVerifier) {
        this._codeVerifier = codeVerifier;
        this.notifyStateUpdated();
    }
    get clientMetadata() {
        return {
            redirect_uris: [this.redirectUri],
            client_name: this.clientName,
        };
    }
    /**
     * Called by the MCP SDK when the user needs to be redirected to authorize.
     * We store the URL so the CLI can open it in the browser.
     */
    redirectToAuthorization(authorizationUrl) {
        this._redirectStartAuthUrl = authorizationUrl;
    }
    async tokens() {
        return this._tokens;
    }
    async saveTokens(tokens) {
        this._tokens = tokens;
        this.notifyStateUpdated();
    }
    /**
     * Get the current state for persistence
     */
    getState() {
        return {
            tokens: this._tokens,
            clientInformation: this._clientInformation,
            codeVerifier: this._codeVerifier,
            serverUrl: this.serverUrl,
        };
    }
    notifyStateUpdated() {
        if (this.onStateUpdated) {
            this.onStateUpdated(this.getState());
        }
    }
}
/**
 * Create a FileOAuthProvider for use with MCP transports.
 * This is the simpler factory function for common use cases.
 */
export function createFileOAuthProvider(options) {
    return new FileOAuthProvider(options);
}
