// unframer/dist/flat-cache-interceptor.js
// FlatCacheStore.ts
import { Writable } from 'node:stream';
import { promises as fs } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import { createHash } from 'node:crypto';
import { logger } from './utils.js';
/**
 * A CacheStore that persists each entry as separate files in os.tmpdir()/.unframer.
 * Each cache entry creates a .json file for metadata and a .bin file for the body.
 */
export class FlatCacheStore {
    cacheDir;
    constructor({ cacheDir }) {
        this.cacheDir = join(tmpdir(), cacheDir);
        logger.log(`using cache dir`, this.cacheDir);
        this.ensureCacheDir();
    }
    async ensureCacheDir() {
        try {
            await fs.mkdir(this.cacheDir, { recursive: true });
        }
        catch {
            // Directory might already exist, ignore error
        }
    }
    getFileHash(key) {
        return createHash('sha256').update(JSON.stringify(key)).digest('hex');
    }
    getFilePaths(key) {
        const hash = this.getFileHash(key);
        return {
            metaPath: join(this.cacheDir, `${hash}.json`),
            bodyPath: join(this.cacheDir, `${hash}.bin`),
        };
    }
    async get(key) {
        try {
            const { metaPath, bodyPath } = this.getFilePaths(key);
            const [metaData, bodyData] = await Promise.all([
                fs.readFile(metaPath, 'utf-8'),
                fs.readFile(bodyPath),
            ]);
            const meta = JSON.parse(metaData);
            return { ...meta, body: bodyData };
        }
        catch {
            return undefined;
        }
    }
    createWriteStream(key, meta) {
        const chunks = [];
        return new Writable({
            write(chunk, _enc, cb) {
                chunks.push(chunk);
                cb();
            },
            final: async (cb) => {
                try {
                    await this.ensureCacheDir();
                    const { metaPath, bodyPath } = this.getFilePaths(key);
                    await Promise.all([
                        fs.writeFile(metaPath, JSON.stringify(meta, null, 2)),
                        fs.writeFile(bodyPath, Buffer.concat(chunks)),
                    ]);
                    cb();
                }
                catch (error) {
                    cb(error);
                }
            },
        });
    }
    async delete(key) {
        try {
            const { metaPath, bodyPath } = this.getFilePaths(key);
            await Promise.all([
                fs.unlink(metaPath).catch(() => { }),
                fs.unlink(bodyPath).catch(() => { }),
            ]);
        }
        catch {
            // Ignore errors when deleting non-existent files
        }
    }
}
